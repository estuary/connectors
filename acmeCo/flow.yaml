collections:
  acmeCo/fake-data/ticks:
    schema: ticks.schema.yaml
    key: [/ts]
  acmeCo/watchdog/messages:
    schema: messages.schema.yaml
    key: [/ts]
    derivation:
      register:
        schema: { type: [string, "null"] }
        initial: null
      typescript:
        module: watchdog.flow.ts
      transform:
        liveStats:
          source: { name: ops/acmeCo/stats }
          shuffle: { key: [/shard/name] }
          update: { lambda: typescript }
        delayedStats:
          readDelay: "1m"
          source: { name: ops/acmeCo/stats }
          shuffle: { key: [/shard/name] }
          publish: { lambda: typescript }

captures:
  acmeCo/fake-data/from-clock:
    endpoint:
      connector:
        image: source-clock:local
        config: source-clock.config.yaml
    bindings:
      - resource:
          stream: tick10s
          syncMode: incremental
        target: acmeCo/fake-data/ticks

materializations:
  acmeCo/watchdog/to-slack:
    endpoint:
      connector:
        image: materialize-slack:local
        config: materialize-slack.config.yaml
    bindings:
      - source: acmeCo/watchdog/messages
        resource:
          channel: "flow-hack-a-thon"
        fields:
          recommended: true

storageMappings: {"": {"stores": [{"bucket": "wgd-test-bucket", "prefix": "flow/", "provider": "GCS"}]}}