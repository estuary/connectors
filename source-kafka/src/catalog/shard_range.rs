use std::ops::RangeInclusive;

use serde::Deserialize;
use serde_with::serde_as;

/// Used to parse the exact format of the Range extension to the Airbyte Spec.
/// This intermediate data structure is then immediately converted into a normal
/// Range.
#[serde_as]
#[derive(Debug, Deserialize, PartialEq)]
pub(super) struct ShardRangeDefinition {
    begin: Hex32,
    end: Hex32,
}

impl From<ShardRangeDefinition> for RangeInclusive<u32> {
    fn from(source: ShardRangeDefinition) -> Self {
        source.begin.into()..=source.end.into()
    }
}

#[serde_as]
#[derive(Debug, Deserialize, PartialEq)]
struct Hex32(#[serde_as(as = "serde_with::hex::Hex")] [u8; 4]);

impl From<Hex32> for u32 {
    fn from(hex: Hex32) -> Self {
        u32::from_be_bytes(hex.0)
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use serde_json::{from_value, json};

    #[test]
    fn shard_range_definition_parsing_test() -> eyre::Result<()> {
        let zero: Hex32 = from_value(json!("00000000"))?;
        let two_five_five: Hex32 = from_value(json!("000000ff"))?;

        assert_eq!(
            ShardRangeDefinition {
                begin: zero,
                end: two_five_five
            },
            from_value(json!({"begin": "00000000", "end": "000000ff"}))?
        );

        Ok(())
    }

    #[test]
    fn hex_32_parsing_test() -> eyre::Result<()> {
        let zero: Hex32 = from_value(json!("00000000"))?;
        let one: Hex32 = from_value(json!("00000001"))?;
        let two_five_five: Hex32 = from_value(json!("000000ff"))?;
        let sixty_four_k: Hex32 = from_value(json!("0000ffff"))?;
        let max: Hex32 = from_value(json!("ffffffff"))?;

        assert_eq!(0, u32::from(zero));
        assert_eq!(1, u32::from(one));
        assert_eq!(255, u32::from(two_five_five));
        assert_eq!((2u32.pow(16) - 1) as u32, u32::from(sixty_four_k));
        assert_eq!(u32::MAX as u32, u32::from(max));

        Ok(())
    }
}
