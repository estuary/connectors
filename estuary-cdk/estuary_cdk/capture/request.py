from pydantic import BaseModel, NonNegativeInt
from typing import Generic

from ..flow import (
    CaptureSpec,
    CollectionSpec,
    ConnectorState,
    ConnectorType,
    EndpointConfig,
    RangeSpec,
    ResourceConfig,
)
from ..pydantic_polyfill import GenericModel


class Spec(BaseModel):
    # TODO(johnny): This shouldn't have a default.
    # This is a temporary accommodation while this fix circulates:
    # https://github.com/estuary/flow/pull/1400
    connectorType: ConnectorType = "IMAGE"


class Discover(GenericModel, Generic[EndpointConfig]):
    connectorType: ConnectorType
    config: EndpointConfig


class ValidateBinding(GenericModel, Generic[ResourceConfig]):
    collection: CollectionSpec
    resourceConfig: ResourceConfig


class Validate(GenericModel, Generic[EndpointConfig, ResourceConfig]):
    name: str
    connectorType: ConnectorType
    config: EndpointConfig
    bindings: list[ValidateBinding[ResourceConfig]] = []


class Apply(GenericModel, Generic[EndpointConfig, ResourceConfig]):
    capture: CaptureSpec[EndpointConfig, ResourceConfig]
    version: str


class Open(GenericModel, Generic[EndpointConfig, ResourceConfig, ConnectorState]):
    capture: CaptureSpec[EndpointConfig, ResourceConfig]
    version: str
    range: RangeSpec
    state: ConnectorState


class Acknowledge(BaseModel):
    checkpoints: NonNegativeInt
