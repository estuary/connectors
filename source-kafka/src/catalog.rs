use highway::{HighwayHash, HighwayHasher};
use proto_flow::flow::RangeSpec;
use serde::{Deserialize, Serialize};

mod shard_range;

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("failed to read the catalog file")]
    File(#[from] std::io::Error),

    #[error("failed to validate connector catalog")]
    Format(#[from] serde_json::Error),

    #[error("cannot subscribe to non-existent stream: {0}")]
    MissingStream(String),
}

pub fn responsible_for_shard(range: &RangeSpec, key: impl Into<ShardKey>) -> bool {
    let hash = key.into().hash();
    range.key_begin <= hash && range.key_end >= hash
}

#[derive(Serialize, Deserialize)]
pub struct Resource {
    pub stream: String,
}

#[derive(Default)]
pub struct ShardKey(HighwayHasher);

impl ShardKey {
    pub fn add_int(mut self, n: impl Into<i64>) -> Self {
        self.0.append(&n.into().to_be_bytes());
        self
    }

    pub fn add_str(mut self, s: &str) -> Self {
        self.0.append(s.as_bytes());
        self
    }

    fn hash(self) -> u32 {
        self.0.finalize64() as u32
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn shard_assignment_test() {
        let range = RangeSpec {
            key_begin: 0,
            key_end: 0x7fffffff,
            r_clock_begin: 0,
            r_clock_end: 0xffffffff,
        };

        let shards = (0..10).map(|n| ShardKey::default().add_int(n));
        let shards_covered = 2;

        assert_eq!(
            shards_covered,
            shards
                .map(|n| responsible_for_shard(&range, n))
                .filter(|b| *b)
                .count()
        );
    }
}
