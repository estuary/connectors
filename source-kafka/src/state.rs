use std::collections::BTreeMap;

use proto_flow::flow::{CaptureSpec, RangeSpec};
use rdkafka::message::BorrowedMessage;
use rdkafka::metadata::Metadata;
use serde::{Deserialize, Serialize};
use tracing::info;

use crate::{catalog::{self, Resource, responsible_for_shard}, connector, kafka};

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("failed to read the state file")]
    File(#[from] std::io::Error),

    #[error("failed to validate connector state")]
    Format(#[from] serde_json::Error),

    #[error("failed to serialize state: {0:?}")]
    Serialization(Checkpoint, serde_json::Error),
}

/// Represents how far into a partition we've already consumed. The `Offset` value
/// stored in a state file is a record of successfully processing a message.
#[derive(Copy, Clone, Debug, Deserialize, PartialEq, Serialize)]
pub enum Offset {
    /// The very beginning of the partition. **Not the same as `UpThrough(0)`**.
    Start,
    /// The very end of the partition. We won't necessarily know the offset number,
    /// but we can use it to avoid reading from the beginning.
    End,
    /// The specific offset we've read. `UpThrough(0)` would mean we've already
    /// consumed the message with offset=0.
    UpThrough(i64),
}

impl Offset {
    /// Returns a new `Offset` representing the next offset to start consuming.
    ///
    /// When we're re-subscribing to a partition, we want to avoid re-reading the
    /// message we last consumed.
    pub fn next(&self) -> Self {
        match self {
            Self::Start => Self::Start,
            Self::End => Self::End,
            Self::UpThrough(n) => Self::UpThrough(*n + 1),
        }
    }
}

impl PartialOrd for Offset {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        use std::cmp::Ordering;

        match (self, other) {
            (Offset::Start, Offset::Start) => Some(Ordering::Equal),
            (Offset::Start, Offset::End) => Some(Ordering::Less),
            (Offset::Start, Offset::UpThrough(-1)) => Some(Ordering::Equal),
            (Offset::Start, Offset::UpThrough(_)) => Some(Ordering::Less),
            (Offset::End, Offset::Start) => Some(Ordering::Greater),
            (Offset::End, Offset::End) => Some(Ordering::Equal),
            (Offset::End, Offset::UpThrough(_)) => Some(Ordering::Greater),
            (Offset::UpThrough(-1), Offset::Start) => Some(Ordering::Equal),
            (Offset::UpThrough(_), Offset::Start) => Some(Ordering::Greater),
            (Offset::UpThrough(_), Offset::End) => Some(Ordering::Less),
            (Offset::UpThrough(l), Offset::UpThrough(r)) => Some(l.cmp(r)),
        }
    }
}

impl From<Offset> for rdkafka::Offset {
    fn from(orig: Offset) -> Self {
        match orig {
            Offset::Start => rdkafka::Offset::Beginning,
            Offset::UpThrough(n) => rdkafka::Offset::Offset(n),
            Offset::End => rdkafka::Offset::End,
        }
    }
}

impl<'m> From<&BorrowedMessage<'m>> for Offset {
    /// When constructing an `Offset` from a Kafka Message directly, we know it's
    /// referencing a offset-value from the start of the partition and that it's
    /// been consumed.
    fn from(msg: &BorrowedMessage) -> Self {
        use rdkafka::Message;

        Offset::UpThrough(msg.offset())
    }
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct Checkpoint {
    pub topic: String,
    pub partition: i32,
    pub offset: Offset,
}

impl Checkpoint {
    pub fn new<O: Into<Offset>>(topic: &str, partition: i32, offset: O) -> Self {
        Self {
            topic: topic.to_owned(),
            partition,
            offset: offset.into(),
        }
    }
}

#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
pub struct CheckpointSet(pub BTreeMap<String, BTreeMap<i32, Offset>>);

impl CheckpointSet {
    pub fn add(&mut self, new_checkpoint: Checkpoint) {
        self.0
            .entry(new_checkpoint.topic)
            .or_insert_with(Default::default)
            .insert(new_checkpoint.partition, new_checkpoint.offset);
    }

    /// Finds the union of the topics in:
    ///   1. The connected Kafka Cluster
    ///   2. The ConfiguredCatalogSpec
    ///   3. The StateFile
    pub fn reconcile_catalog_state(
        metadata: &Metadata,
        capture: &CaptureSpec,
        range: Option<&RangeSpec>,
        loaded_state: &CheckpointSet,
    ) -> Result<CheckpointSet, catalog::Error> {
        let mut reconciled = CheckpointSet::default();

        for binding in capture.bindings.iter() {
            let res: Resource = serde_json::from_str(&binding.resource_config_json)?;
            if let Some(topic) = kafka::find_topic(metadata, &res.stream) {
                for partition in topic.partitions() {
                    if let Some(range) = range {
                        if responsible_for_shard(range, kafka::build_shard_key(topic, partition)) {
                            info!("Responsible for {}/{}: YES", topic.name(), partition.id());

                            let offset = loaded_state
                                .offset_for(topic.name(), partition.id())
                                .unwrap_or(Offset::Start);

                            reconciled.add(Checkpoint::new(topic.name(), partition.id(), offset));
                        } else {
                            info!("Responsible for {}/{}: NO", topic.name(), partition.id());
                        }
                    } else {
                        let offset = loaded_state
                            .offset_for(topic.name(), partition.id())
                            .unwrap_or(Offset::Start);

                        reconciled.add(Checkpoint::new(topic.name(), partition.id(), offset));
                    }
                }
            } else {
                return Err(catalog::Error::MissingStream(
                    res.stream.clone(),
                ));
            }
        }

        Ok(reconciled)
    }

    pub fn offset_for(&self, topic: &str, partition_id: i32) -> Option<Offset> {
        self.0
            .get(topic)
            .and_then(|partitions| partitions.get(&partition_id).map(Clone::clone))
    }

    pub fn iter(&self) -> impl Iterator<Item = Checkpoint> + '_ {
        self.0.iter().flat_map(|(topic, partitions)| {
            partitions
                .iter()
                .map(move |(partition, offset)| Checkpoint::new(topic, *partition, *offset))
        })
    }
}

impl connector::ConnectorConfig for CheckpointSet {
    type Error = Error;

    fn parse(reader: &str) -> Result<Self, Self::Error> {
        let checkpoints = serde_json::from_str(reader)?;

        Ok(checkpoints)
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use connector::ConnectorConfig;

    #[test]
    fn merge_test() {
        let check_offsets = |checkpoints: &CheckpointSet, foo0, foo1, bar0| {
            assert_eq!(checkpoints.0["foo"][&0], foo0);
            assert_eq!(checkpoints.0["foo"][&1], foo1);
            assert_eq!(checkpoints.0["bar"][&0], bar0);
        };

        let update_offset = |checkpoint, new_offset| {
            Checkpoint {
                offset: new_offset,
                ..checkpoint
            }
        };

        let mut checkpoints = CheckpointSet::default();

        let foo0 = Checkpoint::new("foo", 0, Offset::Start);
        let foo1 = Checkpoint::new("foo", 1, Offset::UpThrough(10));
        let bar0 = Checkpoint::new("bar", 0, Offset::Start);

        checkpoints.add(foo0.clone());
        checkpoints.add(foo1.clone());
        checkpoints.add(bar0);

        check_offsets(
            &checkpoints,
            Offset::Start,
            Offset::UpThrough(10),
            Offset::Start,
        );

        let foo0 = update_offset(foo0, Offset::UpThrough(0));
        checkpoints.add(foo0.clone());
        check_offsets(
            &checkpoints,
            Offset::UpThrough(0),
            Offset::UpThrough(10),
            Offset::Start,
        );

        let foo0 = update_offset(foo0, Offset::UpThrough(1));
        checkpoints.add(foo0);
        check_offsets(
            &checkpoints,
            Offset::UpThrough(1),
            Offset::UpThrough(10),
            Offset::Start,
        );

        let foo1 = update_offset(foo1, Offset::UpThrough(11));
        checkpoints.add(foo1);
        check_offsets(
            &checkpoints,
            Offset::UpThrough(1),
            Offset::UpThrough(11),
            Offset::Start,
        );
    }

    #[test]
    fn parse_state_file_test() {
        let input = r#"
                {
                    "test": {
                        "0": { "UpThrough": 100 }
                    }
                }
        "#;

        CheckpointSet::parse(input).expect("to parse");
    }

    #[test]
    fn parse_empty_state_file_test() {
        let input = "{}\n";

        CheckpointSet::parse(input).expect("to parse");
    }
}
