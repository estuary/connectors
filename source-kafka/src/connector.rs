use std::io::Write;

use proto_flow::capture::{Response, request};
use proto_flow::flow::{CaptureSpec, RangeSpec};

pub trait ConnectorConfig: Sized {
    type Error: std::error::Error + Send + Sync + 'static;

    fn parse(reader: &str) -> Result<Self, Self::Error>;
}

pub trait Connector {
    type Config: ConnectorConfig;
    type State: ConnectorConfig;

    fn spec(output: &mut dyn Write) -> eyre::Result<()>;
    fn validate(output: &mut dyn Write, validate: request::Validate) -> eyre::Result<()>;
    fn discover(output: &mut dyn Write, discover: request::Discover) -> eyre::Result<()>;
    fn apply(output: &mut dyn Write, config: Self::Config) -> eyre::Result<()>;
    fn read(
        output: &mut dyn Write,
        config: Self::Config,
        capture: CaptureSpec,
        range: Option<RangeSpec>,
        state: Option<Self::State>,
    ) -> eyre::Result<()>;
}

#[derive(Debug, thiserror::Error)]
pub enum StdoutError {
    #[error("failed to write message to output")]
    Output(#[source] serde_json::Error),

    #[error("failed to emit a newline between messages")]
    OutputNewline(#[source] std::io::Error),
}

pub fn write_message(
    mut output: &mut dyn Write,
    message: Response,
) -> Result<(), StdoutError> {
    serde_json::to_writer(&mut output, &message)
        .map_err(StdoutError::Output)?;

    // Include a newline to break up the document stream.
    writeln!(&mut output).map_err(StdoutError::OutputNewline)?;
    Ok(())
}
