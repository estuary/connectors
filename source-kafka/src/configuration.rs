use std::{fmt::Display, io::Read};

use schemars::JsonSchema;
use serde::{de::Visitor, Deserialize, Serialize};

use crate::connector;

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("failed to read the configuration file")]
    File(#[from] std::io::Error),

    #[error("failed to parse the file as valid json")]
    Parsing(#[from] serde_json::Error),

    #[error("bootstrap servers are required to make the initial connection")]
    NoBootstrapServersGiven,
}

/// # Kafka Connector Configuration
#[derive(Deserialize, Default, JsonSchema, Serialize)]
pub struct Configuration {
    /// The initial servers in the Kafka cluster to initially connect to. The Kafka
    /// client will be informed of the rest of the cluster nodes by connecting to
    /// one of these nodes.
    #[schemars(with = "Vec<String>")]
    pub bootstrap_servers: Vec<BootstrapServer>,

    /// The connection details for authenticating a client connection to Kafka via SASL.
    /// When not provided, the client connection will attempt to use PLAINTEXT
    /// (insecure) protocol. This must only be used in dev/test environments.
    pub authentication: Option<Authentication>,

    /// The TLS connection settings.
    pub tls: TlsSettings,
}

impl Configuration {
    pub fn brokers(&self) -> String {
        self.bootstrap_servers
            .iter()
            .map(|url| url.to_string())
            .collect::<Vec<String>>()
            .join(",")
    }

    pub fn security_protocol(&self) -> &'static str {
        match (&self.authentication, &self.tls) {
            (None, TlsSettings::SystemCertificates) => "SSL",
            (None, TlsSettings::Cleartext) => "PLAINTEXT",
            (Some(_), TlsSettings::SystemCertificates) => "SASL_SSL",
            (Some(_), TlsSettings::Cleartext) => "SASL_PLAINTEXT",
        }
    }
}

impl connector::ConnectorConfig for Configuration {
    type Error = Error;

    fn parse(reader: impl Read) -> Result<Self, Self::Error> {
        let configuration: Configuration = serde_json::from_reader(reader)?;

        if configuration.bootstrap_servers.is_empty() {
            return Err(Error::NoBootstrapServersGiven);
        }

        Ok(configuration)
    }
}

// Note: The Rust `url` crate doesn't like `ip:port` pairs without a scheme.
// Kafka doesn't specify a scheme for the `bootstrap_servers` values, so
// expecting anyone to add one is really odd. Thus, we parse these values
// ourselves.
pub struct BootstrapServer {
    host: String,
    port: u16,
}

impl BootstrapServer {
    pub fn new(host: impl Into<String>, port: u16) -> Self {
        Self {
            host: host.into(),
            port,
        }
    }
}

impl Display for BootstrapServer {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:{}", &self.host, self.port)
    }
}

impl Serialize for BootstrapServer {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&self.to_string())
    }
}

impl<'de> Deserialize<'de> for BootstrapServer {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        deserializer.deserialize_str(BootstrapServerVisitor)
    }
}

struct BootstrapServerVisitor;

impl<'de> Visitor<'de> for BootstrapServerVisitor {
    type Value = BootstrapServer;

    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(formatter, "a host and a port, split by a `:`")
    }

    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
    where
        E: serde::de::Error,
    {
        if let Some((host, port)) = v.split_once(':') {
            if let Ok(port_num) = port.parse::<u16>() {
                Ok(BootstrapServer::new(host, port_num))
            } else {
                Err(E::custom(format!(
                    "expected the port to be a u16. got: `{}`",
                    port
                )))
            }
        } else {
            Err(E::custom(format!("expected to find a colon. got: `{}`", v)))
        }
    }
}

/// The SASL Mechanism describes _how_ to exchange and authenticate
/// clients/servers. For secure communication, TLS is **required** for all
/// supported mechanisms.
///
/// For more information about the Simple Authentication and Security Layer (SASL), see RFC 4422:
/// https://datatracker.ietf.org/doc/html/rfc4422
/// For more information about Salted Challenge Response Authentication
/// Mechanism (SCRAM), see RFC 7677.
/// https://datatracker.ietf.org/doc/html/rfc7677
#[derive(Debug, Deserialize, JsonSchema, Serialize)]
#[serde(rename_all = "SCREAMING-KEBAB-CASE")]
pub enum SaslMechanism {
    /// The username and password are sent to the server in the clear.
    Plain,
    /// SCRAM using SHA-256.
    #[serde(rename = "SCRAM-SHA-256")]
    ScramSha256,
    /// SCRAM using SHA-512.
    #[serde(rename = "SCRAM-SHA-512")]
    ScramSha512,
}

impl Display for SaslMechanism {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SaslMechanism::Plain => write!(f, "PLAIN"),
            SaslMechanism::ScramSha256 => write!(f, "SCRAM-SHA-256"),
            SaslMechanism::ScramSha512 => write!(f, "SCRAM-SHA-512"),
        }
    }
}

/// The information necessary to connect to Kafka.
#[derive(Debug, Deserialize, JsonSchema, Serialize)]
pub struct Authentication {
    pub mechanism: SaslMechanism,
    pub username: String,
    pub password: String,
}

#[derive(Debug, Deserialize, JsonSchema, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum TlsSettings {
    /// Use the TLS certificates found on the system by openssl.
    SystemCertificates,
    /// Disable TLS and use an insecure connection to the Kafka brokers. This
    /// should only be used for dev/test environments.
    Cleartext,
    // TODO: allow the user to specify custom TLS certs, authorities, etc.
    // CustomCertificates(CustomTlsSettings),
}

impl Default for TlsSettings {
    fn default() -> Self {
        TlsSettings::SystemCertificates
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn empty_brokers_test() {
        let config = Configuration::default();
        let brokers = config.brokers();
        assert_eq!("", brokers);
    }

    #[test]
    fn many_brokers_test() {
        let config: Configuration = serde_json::from_str(
            r#"{
            "bootstrap_servers": [
                "localhost:9092",
                "172.22.36.2:9093",
                "localhost:9094"
            ],
            "tls": "system_certificates"
        }"#,
        )
        .expect("to parse the config");

        let brokers = config.brokers();
        assert_eq!("localhost:9092,172.22.36.2:9093,localhost:9094", brokers);
    }

    #[test]
    fn parse_config_file_test() {
        use connector::ConnectorConfig;

        let input = std::io::Cursor::new(
            r#"
        {
            "bootstrap_servers": ["localhost:9093"],
            "tls": "system_certificates"
        }
        "#,
        );

        Configuration::parse(input).expect("to parse");

        let input = std::io::Cursor::new(
            r#"
        {
            "bootstrap_servers": ["localhost:9093"],
            "authentication": {
                "mechanism": "SCRAM-SHA-256",
                "username": "user",
                "password": "password"
            },
            "tls": "cleartext"
        }
        "#,
        );

        Configuration::parse(input).expect("to parse");
    }
}
