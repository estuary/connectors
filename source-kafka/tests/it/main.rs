use std::fs::File;
use std::io::Read;

use proto_flow::capture::request;
use proto_flow::flow::capture_spec;
use proto_flow::flow::CaptureSpec;
use proto_flow::flow::CollectionSpec;
use source_kafka::catalog;
use source_kafka::configuration;
use source_kafka::connector::Connector;
use source_kafka::connector::ConnectorConfig;
use source_kafka::state;
use support::assert_valid_json;
use support::mock_stdout;

use crate::support::parse_from_output;
use crate::support::parse_messages_from_output;

mod support;

#[test]
fn spec_test() {
    let mut stdout = mock_stdout();

    source_kafka::KafkaConnector::spec(&mut stdout).expect("spec command to succeed");

    insta::assert_yaml_snapshot!(parse_from_output(&stdout));
}

#[test]
fn check_test() {
    let mut stdout = mock_stdout();
    let config = include_str!("../test-config.json");
    let req = request::Validate {
        name: "test/source-kafka".to_string(),
        connector_type: capture_spec::ConnectorType::Image as i32,
        config_json: config.to_string(),
        // The connector seems to ignore bindings during validation.
        // That seems wrong, but not something I'm going to try to address at the moment.
        bindings: Vec::new(),
    };

    source_kafka::KafkaConnector::validate(&mut stdout, req).expect("check command to succeed");

    insta::assert_yaml_snapshot!(parse_from_output(&stdout), {
        ".connectionStatus.message" => "{{ NUM_TOPICS_FOUND }}"
    });
}

#[test]
fn discover_test() {
    let mut stdout = mock_stdout();
    let config = include_str!("../test-config.json");
    let req = request::Discover {
        connector_type: capture_spec::ConnectorType::Image as i32,
        config_json: config.to_string(),
    };

    source_kafka::KafkaConnector::discover(&mut stdout, req).expect("discover command to succeed");

    // This is tricky to snapshot. It detects any other topics within the
    // connected Kafka, which will vary by dev machine.
    assert_valid_json(&stdout);
}

#[test]
fn read_simple_catalog_test() {
    let mut stdout = mock_stdout();
    let config = local_config();
    let catalog = local_capture("todo-list");

    source_kafka::KafkaConnector::read(&mut stdout, config, catalog, None, None)
        .expect("read command to succeed");

    let messages = parse_messages_from_output(&stdout);
    // Only look at the last 10 messages. Otherwise the snapshot file gets unwieldy.
    let last_ten = (messages.len() - 10)..;
    insta::assert_yaml_snapshot!(&messages[last_ten], {
        "[].record.emitted_at" => "{{ UNIX_TIMESTAMP }}"
    });
}

#[test]
fn read_resume_from_state_test() {
    let mut stdout = mock_stdout();
    let config = local_config();
    let catalog = local_capture("todo-list");

    let mut state = state::CheckpointSet::default();
    state.add(state::Checkpoint::new(
        "todo-list",
        0,
        state::Offset::UpThrough(37),
    ));
    state.add(state::Checkpoint::new(
        "todo-list",
        2,
        state::Offset::UpThrough(57),
    ));

    source_kafka::KafkaConnector::read(&mut stdout, config, catalog, None, Some(state))
        .expect("read command to succeed");

    insta::assert_yaml_snapshot!(parse_messages_from_output(&stdout), {
        "[].record.emitted_at" => "{{ UNIX_TIMESTAMP }}"
    });
}

fn local_config() -> configuration::Configuration {
    let mut file = File::open("tests/test-config.json").expect("to open test config file");
    let mut buf = String::new();
    file.read_to_string(&mut buf)
        .expect("to read test config file");
    configuration::Configuration::parse(&buf).expect("to parse test config file")
}

fn local_capture(binding_name: &str) -> CaptureSpec {
    let config = include_str!("../test-config.json");
    CaptureSpec {
        name: "capture".to_string(),
        bindings: vec![capture_spec::Binding {
            resource_config_json: serde_json::to_string(&catalog::Resource {
                stream: binding_name.to_string(),
            })
            .unwrap(),
            resource_path: vec![binding_name.to_string()],
            collection: Some(CollectionSpec {
                ..Default::default()
            }),
            backfill: 0,
            state_key: "".to_string(),
        }],
        connector_type: capture_spec::ConnectorType::Image as i32,
        config_json: config.to_string(),
        interval_seconds: 30,
        shard_template: None,
        recovery_log_template: None,
        network_ports: Vec::new(),
    }
}
