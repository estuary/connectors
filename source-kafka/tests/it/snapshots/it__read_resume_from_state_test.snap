---
source: tests/it/main.rs
expression: parse_messages_from_output(&stdout)

---
- record:
    data:
      completed: true
      id: 159
      title: debitis adipisci quibusdam aliquam sed dolore ea praesentium nobis
      userId: 8
    emitted_at: "{{ UNIX_TIMESTAMP }}"
    namespace: Partition 0
    stream: todo-list
  type: RECORD
- state:
    data:
      todo-list:
        "0":
          UpThrough: 38
  type: STATE
- record:
    data:
      completed: false
      id: 160
      title: et praesentium aliquam est
      userId: 8
    emitted_at: "{{ UNIX_TIMESTAMP }}"
    namespace: Partition 0
    stream: todo-list
  type: RECORD
- state:
    data:
      todo-list:
        "0":
          UpThrough: 39
  type: STATE
- record:
    data:
      completed: true
      id: 199
      title: numquam repellendus a magnam
      userId: 10
    emitted_at: "{{ UNIX_TIMESTAMP }}"
    namespace: Partition 2
    stream: todo-list
  type: RECORD
- state:
    data:
      todo-list:
        "2":
          UpThrough: 58
  type: STATE
- record:
    data:
      completed: false
      id: 200
      title: ipsam aperiam voluptates qui
      userId: 10
    emitted_at: "{{ UNIX_TIMESTAMP }}"
    namespace: Partition 2
    stream: todo-list
  type: RECORD
- state:
    data:
      todo-list:
        "2":
          UpThrough: 59
  type: STATE

